File: src/main/java/ai/aimachineserver/domain/games/soccer/GameSoccer.java
package ai.aimachineserver.domain.games.soccer;

import ai.aimachineserver.domain.games.AbstractGame;
import ai.aimachineserver.domain.games.KeyValue;
import org.json.JSONObject;
import org.springframework.web.socket.WebSocketSession;

import static ai.aimachineserver.domain.games.soccer.TurnResultSoccer.TURN_ONGOING;

public class GameSoccer extends AbstractGame {
    private static final PlayerSoccer playerStub = new PlayerSoccerHuman("");
    private final BoardSoccer board;
    private final JudgeSoccer judge;

    private PlayerSoccer player1 = playerStub;
    private PlayerSoccer player2 = playerStub;
    private PlayerSoccer currentPlayer = player1;
    private TurnResultSoccer turnResult = TURN_ONGOING;

    public GameSoccer(BoardSoccer board, JudgeSoccer judge, String gameName) {
        super(gameName);
        this.board = board;
        this.judge = judge;
    }

    @Override
    public void onPlayerJoinedGame(WebSocketSession session) {
        this.playerSessions.add(session);
        if (player1 == playerStub) {
            player1 = new PlayerSoccerHuman(session.getId());
            currentPlayer = player1;
        } else {
            player2 = new PlayerSoccerHuman(session.getId());
            broadcastMessage(new KeyValue("eventType", "current_player"), new KeyValue("eventMessage", currentPlayer.getName()));
            JSONObject json = new JSONObject();
            json.put("player1", player1.getName());
            json.put("player2", player2.getName());
            String data = json.toString();
            broadcastMessage(new KeyValue("eventType", "players_in_game"), new KeyValue("eventMessage", data));
            broadcastMessage(new KeyValue("eventType", "game_started"), new KeyValue("eventMessage", "game starting"));
        }
        int playersCount = playerSessions.size();
        broadcastMessage(new KeyValue("eventType", "players_count"), new KeyValue("eventMessage", "$playersCount"));
        String message = playersCount == 1 ? "Waiting for opponent" : "Game has started";
        broadcastMessage(new KeyValue("eventType", "server_message"), new KeyValue("eventMessage", message));
    }

    @Override
    public void onFieldClicked(int rowIndex, int colIndex) {
        if (board.isFieldAvailable(rowIndex, colIndex)) {
            System.out.printf("%s clicked [row, col]: [%d, %d]", currentPlayer == player1 ? "player1" : "player2", rowIndex, colIndex);
            JSONObject json = new JSONObject();
            json.put("rowIndex", rowIndex);
            json.put("colIndex", colIndex);
            String data = json.toString();
            broadcastMessage(new KeyValue("eventType", "new_move_to_mark"), new KeyValue("eventMessage", data));
            currentPlayer.makeMove(board, rowIndex, colIndex);
            turnResult = judge.announceTurnResult();
            switch (turnResult) {
                case TURN_OVER:
                    currentPlayer = assignPlayer();
                    broadcastMessage(
                            new KeyValue("eventType", "current_player"),
                            new KeyValue("eventMessage", currentPlayer.getName())
                    );
                    break;
                case TURN_ONGOING:
                    broadcastMessage(
                            new KeyValue("eventType", "current_player"),
                            new KeyValue("eventMessage", currentPlayer.getName())
                    );
                    break;
                default:
                    String resultMessage = getEndgameMessage();
                    broadcastMessage(new KeyValue("eventType", "game_ended"), new KeyValue("eventMessage", resultMessage));
                    System.out.println(resultMessage);
                    broadcastMessage(
                            new KeyValue("eventType", "current_player"),
                            new KeyValue("eventMessage", "none")
                    );
            }
        }
    }

    private PlayerSoccer assignPlayer() {
        if (currentPlayer == player1) {
            return player2;
        } else {
            return player1;
        }
    }

    private String getEndgameMessage() {
        switch (turnResult) {
            case WIN:
                return formatMessage(currentPlayer == player1);
            case LOSE:
                return formatMessage(currentPlayer != player1);
            default:
                return "Unexpected state, everyone is a winner";
        }
    }

    private String formatMessage(Boolean condition) {
        return String.format("%s won", condition ? "player1 " + player1.getName() : "player2 " + player2.getName());
    }

    @Override
    public void onDisconnect(WebSocketSession session) {
        try {
            playerSessions.remove(session);
            broadcastMessage(
                    new KeyValue("eventType", "server_message"),
                    new KeyValue("eventMessage", "Player disconnected")
            );
            broadcastMessage(
                    new KeyValue("eventType", "server_message"),
                    new KeyValue("eventMessage", "${playerSessions.count()} players in game")
            );
        } catch (NullPointerException e) {
            System.out.println("No client ref. Both clients already disconnected. Cannot broadcast message to missing socket");
        } catch (IllegalStateException e) {
            System.out.println("Illegal state. Both clients already disconnected. Cannot broadcast message to missing socket");
        }
    }
}

File: src/main/java/ai/aimachineserver/domain/games/soccer/GameSoccer.kt
package ai.aimachineserver.domain.games.soccer

import ai.aimachineserver.domain.games.AbstractGame
import org.json.JSONObject
import org.springframework.web.socket.WebSocketSession

class GameSoccer(
    private val board: BoardSoccer = BoardSoccer(),
    private val judge: JudgeSoccer = JudgeSoccer(board),
    override val gameName: String
) : AbstractGame() {
    private companion object {
        val playerStub = PlayerSoccerHuman("")
    }

    private var player1 = playerStub
    private var player2 = playerStub
    private var currentPlayer = player1
    private var turnResult = TurnResultSoccer.TURN_ONGOING
    override val playerSessions = mutableListOf<WebSocketSession>()

    override fun onPlayerJoinedGame(session: WebSocketSession) {
        playerSessions.add(session)
        if (player1 == playerStub) {
            player1 = PlayerSoccerHuman(session.id)
            currentPlayer = player1
        } else {
            player2 = PlayerSoccerHuman(session.id)
            broadcastMessage("eventType" to "current_player", "eventMessage" to currentPlayer.name)
            val data = JSONObject()
                .put("player1", player1.name)
                .put("player2", player2.name)
                .toString()
            broadcastMessage("eventType" to "players_in_game", "eventMessage" to data)
            broadcastMessage("eventType" to "game_started", "eventMessage" to "game starting")
        }
        val playersCount = playerSessions.count()
        broadcastMessage("eventType" to "players_count", "eventMessage" to "$playersCount")
        val message = if (playersCount == 1) "Waiting for opponent" else "Game has started"
        broadcastMessage("eventType" to "server_message", "eventMessage" to message)
    }

    override fun onFieldClicked(rowIndex: Int, colIndex: Int) {
        if (board.isFieldAvailable(rowIndex, colIndex)) {
            println("${if (currentPlayer == player1) "player1" else "player2"} clicked [row, col]: [$rowIndex, $colIndex]")
            val data = JSONObject()
                .put("rowIndex", rowIndex)
                .put("colIndex", colIndex)
                .toString()
            broadcastMessage("eventType" to "new_move_to_mark", "eventMessage" to data)
            currentPlayer.makeMove(board, rowIndex, colIndex)
            turnResult = judge.announceTurnResult()
            when (turnResult) {
                TurnResultSoccer.TURN_OVER -> {
                    currentPlayer = assignPlayer()
                    broadcastMessage(
                        "eventType" to "current_player",
                        "eventMessage" to currentPlayer.name
                    )
                }
                TurnResultSoccer.TURN_ONGOING -> {
                    broadcastMessage(
                        "eventType" to "current_player",
                        "eventMessage" to currentPlayer.name
                    )
                }
                else -> {
                    val resultMessage = getEndgameMessage()
                    broadcastMessage("eventType" to "game_ended", "eventMessage" to resultMessage)
                    println(resultMessage)
                    broadcastMessage(
                        "eventType" to "current_player",
                        "eventMessage" to "none"
                    )
                }
            }
        }
    }

    private fun assignPlayer() = if (currentPlayer == player1) {
        player2
    } else {
        player1
    }

    private fun getEndgameMessage() = when (turnResult) {
        TurnResultSoccer.WIN -> formatMessage(currentPlayer == player1)
        TurnResultSoccer.LOSE -> formatMessage(currentPlayer != player1)
        else -> "Unexpected state, everyone is a winner"
    }

    private fun formatMessage(condition: Boolean) =
        "${(if (condition) "player 1 (${player1.name})" else "player 2 (${player2.name})")} won"

    override fun onDisconnect(session: WebSocketSession) {
        try {
            playerSessions.remove(session)
            broadcastMessage(
                "eventType" to "server_message",
                "eventMessage" to "Player disconnected"
            )
            broadcastMessage(
                "eventType" to "server_message",
                "eventMessage" to "${playerSessions.count()} players in game"
            )
        } catch (e: NullPointerException) {
            println("No client ref. Both clients already disconnected. Cannot broadcast message to missing socket")
        } catch (e: IllegalStateException) {
            println("Illegal state. Both clients already disconnected. Cannot broadcast message to missing socket")
        }
    }
}
